Doc/using/cmdline.rst

```diff
@@ -994,3 +994,101 @@ Debug-mode variables
    Need Python configured with the :option:`--with-trace-refs` build option.
 
    .. versionadded:: 3.11
+
+.. _using-on-finding-modules:
+
+Finding modules
+---------------
+
+A module search path is initialized when Python starts. This module search path
+may be accessed at :data:`sys.path`.
+
+The first entry in the module search path is the directory that contains the
+input script, if there is one. Otherwise, the first entry is the current
+directory, which is the case when executing the interactive shell, a ``-c``
+command, or ``-m`` module.
+
+The :envvar:`PYTHONPATH` environment variable is often used to add directories
+to the search path. If this environment variable is found then the contents are
+added to the module search path.
+
+The next items added are the directories containing standard Python modules as
+well as any shared libraries (DLL/dylib/so) that these modules depend on. The
+directory with the platform-independent Python modules is called ``prefix``.
+The directory with the shared libraries is called ``exec_prefix``.
+
+The :envvar:`PYTHONHOME` environment variable may be used to set the ``prefix``
+and ``exec_prefix`` locations. Otherwise these directories are found by using
+the Python executable as a starting point and then looking for various 'landmark'
+files and directories. Note that any symbolic links are followed so the real
+Python executable location is used as the search starting point. On MacOS, the
+:envvar:`PYTHONEXECUTABLE` environment variable may be used to set the Python
+executable location. The Python executable location is called ``home``.
+
+Once ``home`` is determined, the ``prefix`` directory is found by first looking
+for :file:`python{majorversion}{minorversion}.zip` (``python311.zip``). On Windows
+the zip archive is searched for in ``home`` and on Unix the archive is expected
+to be in :file:`lib`. Note that the expected zip archive location is added to the
+module search path even if the archive does not exist. If no archive was found,
+Python on Windows will continue the search for ``prefix`` by looking for :file:`Lib\\os.py`
+or :file:`Lib\\os.pyc`. Python on Unix will look for :file:`lib/python{majorversion}.{minorversion}/os.py`
+(``lib/python3.11/os.py``) or :file:`lib/python{majorversion}.{minorversion}/os.pyc` (``lib/python3.11/os.pyc``).
```

Comment:
Is that true? On Unix systems these days, the .pyc files are supposed to be in `lib/python3.n/__pycache__/` and have a version-specific and opt-level specific suffix. For example, on a current macOS framework install:

```
$ cd /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10
$ ls *.pyc
ls: *.pyc: No such file or directory
$ ls __pycache__/
__future__.cpython-310.opt-1.pyc                    nntplib.cpython-310.opt-1.pyc
__future__.cpython-310.pyc                          nntplib.cpython-310.pyc
__phello__.foo.cpython-310.opt-1.pyc                ntpath.cpython-310.opt-1.pyc
[...]
```

Reply:
Well spotted! I think you are correct, this should be updated in getpath.py I believe, @zooba what do you think?

At the moment line 170 in getpath.py has:
```
if os_name == 'posix' or os_name == 'darwin':
    BUILDDIR_TXT = 'pybuilddir.txt'
    BUILD_LANDMARK = 'Modules/Setup.local'
    DEFAULT_PROGRAM_NAME = f'python{VERSION_MAJOR}'
    STDLIB_SUBDIR = f'{platlibdir}/python{VERSION_MAJOR}.{VERSION_MINOR}'
    STDLIB_LANDMARKS = [f'{STDLIB_SUBDIR}/os.py', f'{STDLIB_SUBDIR}/os.pyc']
    PLATSTDLIB_LANDMARK = f'{platlibdir}/python{VERSION_MAJOR}.{VERSION_MINOR}/lib-dynload'
    BUILDSTDLIB_LANDMARKS = ['Lib/os.py']
    VENV_LANDMARK = 'pyvenv.cfg'
    ZIP_LANDMARK = f'{platlibdir}/python{VERSION_MAJOR}{VERSION_MINOR}.zip'
    DELIM = ':'
    SEP = '/'

elif os_name == 'nt':
    BUILDDIR_TXT = 'pybuilddir.txt'
    BUILD_LANDMARK = f'{VPATH}\\Modules\\Setup.local'
    DEFAULT_PROGRAM_NAME = f'python'
    STDLIB_SUBDIR = 'Lib'
    STDLIB_LANDMARKS = [f'{STDLIB_SUBDIR}\\os.py', f'{STDLIB_SUBDIR}\\os.pyc']
    PLATSTDLIB_LANDMARK = f'{platlibdir}'
    BUILDSTDLIB_LANDMARKS = ['Lib\\os.py']
    VENV_LANDMARK = 'pyvenv.cfg'
    ZIP_LANDMARK = f'python{VERSION_MAJOR}{VERSION_MINOR}{PYDEBUGEXT or ""}.zip'
    WINREG_KEY = f'SOFTWARE\\Python\\PythonCore\\{PYWINVER}\\PythonPath'
    DELIM = ';'
    SEP = '\\'

```
Both the Windows and Unix os.pyc STDLIB_LANDMARKS should be updated/removed.

Reply:
Created https://bugs.python.org/issue46909 for this

Reply:
Looking for `f'{STDLIB_SUBDIR}\\os.pyc'` is even more historical now. Unless the standard library is manually precompiled, which is optional when installing, Python 3.11 won't have an "os[.\*].pyc" file in any directory. The `os` module is frozen in 3.11, so there's no reason to distribute or cache it as a PYC file.

Reply:
@zooba closed https://bugs.python.org/issue46909 as the searching for .pyc is used. Probably best to leave it out of the the documentation though.

Reply:
> The os module is frozen in 3.11, so there's no reason to distribute or cache it as a PYC file.

This is a good point, we probably need to pick a different landmark (perhaps `site.py`?).

I do think detecting the precompiled PYC is correct here. Precompiling on install generates the `__pycache__` directories, but precompiling manually for an embedded runtime is totally legitimate (though I believe more performant if you also zip up the .pycs and put the ZIP file somewhere discoverable, which is why the ZIP detection overrides these landmarks).

Reply:
> This is a good point, we probably need to pick a different landmark (perhaps `site.py`?).

site.py is also frozen. See [Tools/scripts/freeze_modules.py](https://github.com/python/cpython/blob/v3.11.0a6/Tools/scripts/freeze_modules.py#L35-L84). Maybe add an empty landmark file that includes the Python version in its name, e.g. `__py_stdlib_cpython_3_11`.

Reply:
> 

Steve, to be clear, do you think the docs should mention searching for os.pyc or not?